tests/unit/s2n_certificate_signatures_test.c

```diff
@@ -0,0 +1,247 @@
+            EXPECT_NOT_NULL(cert = PEM_read_bio_X509(certBio, NULL, NULL, NULL));
+
+            EXPECT_OK(s2n_is_certificate_sig_scheme_supported(conn, cert, &pss_certificate_signature_preferences, &out));
+            EXPECT_TRUE(out);
+
+            EXPECT_SUCCESS(BIO_free(certBio));
+            X509_free(cert);
+        }
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+    /* s2n_validate_certificate_signature */
+    {
+        /* Connection using the default security policy ignores SHA-1 signatures in certificates */
+        {
+            struct s2n_connection *conn;
+            struct s2n_config *config = s2n_config_new();
+
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default"));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            EXPECT_SUCCESS(s2n_read_test_pem(S2N_RSA_2048_PKCS1_CERT_CHAIN, (char *)cert_file, S2N_MAX_TEST_PEM_SIZE));
+            certLen = strlen((const char*)cert_file);
+
+            /* Read the test certificates into an Openssl X509 struct */
+            EXPECT_NOT_NULL(certBio = BIO_new(BIO_s_mem()));
+            EXPECT_SUCCESS(BIO_write(certBio, cert_file, certLen));
+            EXPECT_NOT_NULL(cert = PEM_read_bio_X509(certBio, NULL, NULL, NULL));
+
+            EXPECT_OK(s2n_validate_certificate_signature(conn, cert));
+
+            EXPECT_SUCCESS(s2n_config_free(config));
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(BIO_free(certBio));
+            X509_free(cert);
+        }
+
+        /* Connection using the default_tls13 security policy does not validate SHA-1 signatures in certificates */
+        {
+            struct s2n_connection *conn;
+            struct s2n_config *config = s2n_config_new();
+
+            EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+            EXPECT_SUCCESS(s2n_config_set_cipher_preferences(config, "default_tls13"));
+            EXPECT_SUCCESS(s2n_connection_set_config(conn, config));
+
+            EXPECT_SUCCESS(s2n_read_test_pem(S2N_RSA_2048_PKCS1_CERT_CHAIN, (char *)cert_file, S2N_MAX_TEST_PEM_SIZE));
+            certLen = strlen((const char*)cert_file);
+
+            /* Read the test certificates into an Openssl X509 struct */
+            EXPECT_NOT_NULL(certBio = BIO_new(BIO_s_mem()));
+            EXPECT_SUCCESS(BIO_write(certBio, cert_file, certLen));
+            EXPECT_NOT_NULL(cert = PEM_read_bio_X509(certBio, NULL, NULL, NULL));
+
+            EXPECT_ERROR_WITH_ERRNO(s2n_validate_certificate_signature(conn, cert), S2N_ERR_CERT_UNTRUSTED);
+
+            EXPECT_SUCCESS(s2n_config_free(config));
+            EXPECT_SUCCESS(s2n_connection_free(conn));
+            EXPECT_SUCCESS(BIO_free(certBio));
+            X509_free(cert);
+        }
+
+        /* Connection using the default_tls13 security policy ignores a SHA-1 signature on a root certificate */
```

Comment:
Isn't this already covered on your more detailed s2n_validate_certificate_signature tests? I'd assume these tests are just intended to verify we're either doing any signature scheme validation / doing no signature scheme validation for common policies.

Reply:
Are you talking about the s2n_x509_validator_tests? I'm not sure what other tests you're referring to. Since the s2n_validate_certificate_signature function checks to see if the cert is self-signed or not, I thought we should include this test. But if you think it's redundant, I can remove it.